File: quests/quest00/ex03.py
"""ex03 - a file named `file` holding the word content. 2 points.

The real gandalf grades what was pushed, so committing is implicit. Offline we
grade the working tree and add an advisory line when the file is not tracked by
git yet, which is the part students actually forget.
"""

import os
import subprocess

FILE = "file"
WANTED = b"content\n"


def _git(origin, *args):
    try:
        proc = subprocess.run(("git",) + args, cwd=origin, timeout=10,
                              stdout=subprocess.PIPE, stderr=subprocess.PIPE)
        return proc.returncode, proc.stdout
    except (OSError, subprocess.TimeoutExpired):
        return 1, b""


def run(g):
    exists = g.check("%s exists and is a regular file" % FILE,
                     g.is_file(FILE),
                     "" if g.exists(FILE) else "no such file in this directory")

    data = g.read_bytes(FILE) if exists else None
    ok = data in (WANTED, b"content")
    g.check('%s contains "content"' % FILE, ok,
            "" if ok else "got %r, expected %r" % (data, WANTED))

    if not exists:
        return

    if not os.path.isdir(os.path.join(g.origin, ".git")) and \
            _git(g.origin, "rev-parse", "--git-dir")[0] != 0:
        g.note("this directory is not inside a git repository")
        return

    code, _ = _git(g.origin, "ls-files", "--error-unmatch", FILE)
    if code != 0:
        g.note("%s is not tracked by git yet: git add %s && git commit -m '...' && git push"
               % (FILE, FILE))
        return

    code, out = _git(g.origin, "status", "--porcelain", "--", FILE)
    if code == 0 and out.strip():
        g.note("%s has uncommitted changes: commit and push them" % FILE)

File: quests/quest00/ex03.js
"use strict";
// A file named `file` holding the word content. 2 points.
//
// The real gandalf grades what was pushed, so committing is implicit. Offline
// we grade the working tree and add an advisory when the file isn't tracked
// by git yet, which is the part students actually forget.

const fs = require("fs");
const path = require("path");
const { spawnSync } = require("child_process");

const FILE = "file";
const WANTED = "content\n";

function git(origin, ...args) {
  const result = spawnSync("git", args, { cwd: origin, timeout: 10000 });
  return { code: result.status === null ? 1 : result.status, stdout: (result.stdout || Buffer.alloc(0)).toString() };
}

exports.run = function (g) {
  const exists = g.check(
    `${FILE} exists and is a regular file`,
    g.isFile(FILE),
    g.exists(FILE) ? "" : "no such file in this directory"
  );

  const data = exists ? g.readText(FILE) : null;
  const ok = data === WANTED || data === "content";
  g.check(
    `${FILE} contains "content"`,
    ok,
    ok ? "" : `got ${JSON.stringify(data)}, expected ${JSON.stringify(WANTED)}`
  );

  if (!exists) return;

  if (!fs.existsSync(path.join(g.origin, ".git")) && git(g.origin, "rev-parse", "--git-dir").code !== 0) {
    g.note("this directory is not inside a git repository");
    return;
  }

  if (git(g.origin, "ls-files", "--error-unmatch", FILE).code !== 0) {
    g.note(`${FILE} is not tracked by git yet: git add ${FILE} && git commit -m '...' && git push`);
    return;
  }

  const status = git(g.origin, "status", "--porcelain", "--", FILE);
  if (status.code === 0 && status.stdout.trim()) {
    g.note(`${FILE} has uncommitted changes: commit and push them`);
  }
};
